[Bug] CGO target headers not added to search paths when imported as external repository

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of rules_go are you using?
0.31.0

<!-- Check io_bazel_rules_go in WORKSPACE if you're not sure -->

### What version of gazelle are you using?
0.24.0

<!-- Check bazel_gazelle in WORKSPACE if you're not sure -->

### What version of Bazel are you using?
5.1.0

<!-- Run "bazel version" to find out -->

### Does this issue reproduce with the latest releases of all the above?
Yes, technically (I can't use Gazelle 0.25.0 due to [this issue](https://github.com/bazelbuild/bazel-gazelle/issues/1217))


### What operating system and processor architecture are you using?
macOS Monterey 12.3.1 (21E258)
2.7 GHz Quad-Core Intel Core i7


### Any other potentially useful information about your toolchain?
Using Go 1.18

<!-- C/C++ compiler, custom CROSSTOOL, remote execution? -->

### What did you do?
I'm currently trying to import a Bazel project that utilizes CGO + JNI (available [here](https://github.com/theopendictionary/odict/tree/nickersoft/bazel5)) into another repository (available [here](https://github.com/nickersoft/go-bazel-repro)).

<!--
If possible, provide a minimal recipe for reproducing the error.
-->

### What did you expect to see?
I expected it to work as usual.


### What did you see instead?
While the [original repo](https://github.com/theopendictionary/odict/tree/nickersoft/bazel5) compiles fine, trying to import it into another project does not add the correct header search paths to the command line, despite the JNI target (`//java/main/cpp`) directly requiring the CGO target (`//bridge`).

The CGO target inside `//bridge` produces a `bridge.h` file that is imported as `bridge/bridge.h` in the [odict.cpp](https://github.com/TheOpenDictionary/odict/blob/master/java/main/cpp/odict.cpp) file. Inside the original repo, the `bazel build java` command executes the following in the sandbox:

```
cd /private/var/tmp/_bazel_tjnickerson/1bdab46624ea7f4407b2d453e6808fb8/sandbox/darwin-sandbox/232/execroot/odict && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=12.3 \
    DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer \
    PATH=/Users/tjnickerson/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.1.1-darwin-x86_64/bin:/Users/tjnickerson/.sdkman/candidates/java/current/bin:/usr/local/opt/ruby@2.7/bin:/Users/tjnickerson/Library/Caches/fnm_multishells/9481_1649548540226/bin:/usr/local/anaconda3/bin/:/Users/tjnickerson/go/bin:/Users/tjnickerson/base/phacility/arcanist/bin:/Users/tjnickerson/.yarn/bin:/Users/tjnickerson/.config/yarn/global/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/tjnickerson/.rvm/bin \
    SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk \
    TMPDIR=/var/folders/0z/yzv926zn0fzbtjjw3p_w0s2r0000gn/T/ \
    XCODE_VERSION_OVERRIDE=13.3.0.13E113 \
    ZERO_AR_DATE=1 \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_tjnickerson/1bdab46624ea7f4407b2d453e6808fb8/sandbox/darwin-sandbox/232/sandbox.sb /var/tmp/_bazel_tjnickerson/install/7a9606eb8f092191f2180bf868dc1b49/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/wrapped_clang_pp '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -O0 -DDEBUG '-std=c++11' 'DEBUG_PREFIX_MAP_PWD=.' -iquote . -iquote bazel-out/darwin-fastbuild/bin -Ibazel-out/darwin-fastbuild-ST-97ca104753ae/bin -isystem tools/jni -isystem bazel-out/darwin-fastbuild/bin/tools/jni -MD -MF bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.o' -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/MacOSX.platform/Developer/Library/Frameworks '-mmacosx-version-min=12.3' -no-canonical-prefixes -pthread -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -target x86_64-apple-macosx -c java/main/cpp/odict.cpp -o bazel-out/darwin-fastbuild/bin/java/main/cpp/_objs/odict-lib/odict.o
```

which includes the following header search paths:

- `.`
- `bazel-out/darwin-fastbuild/bin` 
- `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin` (`bridge/bridge.h` lives here, so the import works as expected)
-  `tools/jni` 
- `bazel-out/darwin-fastbuild/bin/tools/jni`

yet running `bazel build main` on my [reproduction repo](https://github.com/nickersoft/go-bazel-repro) runs this in the sandbox:

```
cd /private/var/tmp/_bazel_tjnickerson/8b9d21129a210b6f322b695a12a00caf/sandbox/darwin-sandbox/142/execroot/__main__ && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=12.3 \
    DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer \
    PATH=/Users/tjnickerson/Library/Caches/bazelisk/downloads/bazelbuild/bazel-5.1.1-darwin-x86_64/bin:/Users/tjnickerson/.sdkman/candidates/java/current/bin:/usr/local/opt/ruby@2.7/bin:/Users/tjnickerson/Library/Caches/fnm_multishells/14807_1649549587978/bin:/usr/local/anaconda3/bin/:/Users/tjnickerson/go/bin:/Users/tjnickerson/base/phacility/arcanist/bin:/Users/tjnickerson/.yarn/bin:/Users/tjnickerson/.config/yarn/global/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/tjnickerson/.rvm/bin \
    SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk \
    TMPDIR=/var/folders/0z/yzv926zn0fzbtjjw3p_w0s2r0000gn/T/ \
    XCODE_VERSION_OVERRIDE=13.3.0.13E113 \
    ZERO_AR_DATE=1 \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_tjnickerson/8b9d21129a210b6f322b695a12a00caf/sandbox/darwin-sandbox/142/sandbox.sb /var/tmp/_bazel_tjnickerson/install/7a9606eb8f092191f2180bf868dc1b49/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/wrapped_clang_pp '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -O0 -DDEBUG '-std=c++11' 'DEBUG_PREFIX_MAP_PWD=.' -iquote external/odict -iquote bazel-out/darwin-fastbuild/bin/external/odict -Ibazel-out/darwin-fastbuild-ST-97ca104753ae/bin -isystem external/odict/tools/jni -isystem bazel-out/darwin-fastbuild/bin/external/odict/tools/jni -MD -MF bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.o' -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/MacOSX.platform/Developer/Library/Frameworks '-mmacosx-version-min=12.3' -no-canonical-prefixes -pthread -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -target x86_64-apple-macosx -c external/odict/java/main/cpp/odict.cpp -o bazel-out/darwin-fastbuild/bin/external/odict/java/main/cpp/_objs/odict-lib/odict.o
```

which uses the following header search paths:

- `external/odict` 
- `bazel-out/darwin-fastbuild/bin/external/odict` (only includes `java`, `schema`, and `tools` – would expect it to contain `bridge` too seeing it's required by the build)
- `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin` (`bridge.h` now technically lives under `bazel-out/darwin-fastbuild-ST-97ca104753ae/bin/external/odict/bridge/bridge.h`, and I have no way of accessing it without changing the original import to `#include "external/odict/bridge/bridge.h"`, which would break the build in the original repo)
- `external/odict/tools/jni`
- `bazel-out/darwin-fastbuild/bin/external/odict/tools/jni`

and throws the error:

```
external/odict/java/main/cpp/odict.cpp:4:10: fatal error: 'bridge/bridge.h' file not found
#include "bridge/bridge.h"
         ^~~~~~~~~~~~~~~~~
1 error generated.
```

This error seems to have cropped up out of nowhere, but I'm wondering if it might be the CGO target's failure to notify Bazel of its generated header or something? Also, if this is a larger-scale Bazel issue, I'm happy to report it to another repo, though I thought seeing the header is originating from CGO this would be a good place to start. Any help here would be greatly appreciated!

**TLDR:** to reproduce, clone [this repo](https://github.com/nickersoft/go-bazel-repro) and run `bazel build main` using Bazel 5.1.0. Should error.

--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -46,6 +46,19 @@ load(
 
 _EMPTY_DEPSET = depset([])
 
+def _include_path(hdr):
+    if not hdr.root.path:
+        fail("Expected hdr to be a generated file, got source file: " + hdr.path)
+
+    root_relative_path = hdr.path[len(hdr.root.path + "/"):]
+    if not root_relative_path.startswith("external/"):
+        return hdr.root.path
+
+    # All headers should be includeable via a path relative to their repository
+    # root, regardless of whether the repository is external or not. If it is,
+    # we thus need to append "external/<external repo name>" to the path.
+    return "/".join([hdr.root.path] + root_relative_path.split("/")[0:2])
+
 def new_cc_import(
         go,
         hdrs = _EMPTY_DEPSET,
@@ -62,7 +75,7 @@ def new_cc_import(
             defines = defines,
             local_defines = local_defines,
             headers = hdrs,
-            includes = depset([hdr.root.path for hdr in hdrs.to_list()]),
+            includes = depset([_include_path(hdr) for hdr in hdrs.to_list()]),
         ),
         linking_context = cc_common.create_linking_context(
             linker_inputs = depset([

--- a/tests/core/cgo/BUILD.bazel
+++ b/tests/core/cgo/BUILD.bazel
@@ -1,4 +1,5 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")
+load("@io_bazel_rules_go//go/tools/bazel_testing:def.bzl", "go_bazel_test")
 
 go_test(
     name = "opts_test",
@@ -407,4 +408,9 @@ cc_library(
     name = "split_import_c",
     srcs = ["split_import_c.c"],
     hdrs = ["split_import_c.h"],
+)
+
+go_bazel_test(
+    name = "external_includes_test",
+    srcs = ["external_includes_test.go"],
 )

--- a/tests/core/cgo/external_includes_test.go
+++ b/tests/core/cgo/external_includes_test.go
@@ -1,0 +1,85 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external_includes_test
+
+import (
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- other_repo/WORKSPACE --
+-- other_repo/cc/BUILD.bazel --
+cc_binary(
+    name = "main",
+    srcs = ["main.c"],
+    deps = ["//cgo"],
+)
+-- other_repo/cc/main.c --
+#include "cgo/cgo.h"
+
+int main() {}
+-- other_repo/cgo/BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
+
+go_binary(
+    name = "cgo",
+    embed = [":cgo_lib"],
+    importpath = "example.com/rules_go/cgo",
+    linkmode = "c-archive",
+    visibility = ["//visibility:public"],
+)
+
+go_library(
+    name = "cgo_lib",
+    srcs = ["cgo.go"],
+    cgo = True,
+    importpath = "example.com/rules_go/cgo",
+    visibility = ["//visibility:private"],
+)
+-- other_repo/cgo/cgo.go --
+package main
+
+import "C"
+
+//export HelloCgo
+func HelloCgo() {}
+
+func main() {}
+`,
+	WorkspaceSuffix: `
+local_repository(
+    name = "other_repo",
+    path = "other_repo",
+)
+`,
+	})
+}
+
+func TestExternalIncludes(t *testing.T) {
+	t.Run("default", func(t *testing.T) {
+		if err := bazel_testing.RunBazel("build", "@other_repo//cc:main"); err != nil {
+			t.Fatalf("Did not expect error:\n%+v", err)
+		}
+	})
+	t.Run("experimental_sibling_repository_layout", func(t *testing.T) {
+		if err := bazel_testing.RunBazel("build", "--experimental_sibling_repository_layout", "@other_repo//cc:main"); err != nil {
+			t.Fatalf("Did not expect error:\n%+v", err)
+		}
+	})
+}

